service/raft/raft_address_map.hh

```diff
@@ -205,12 +291,22 @@ public:
     // This means that we cannot remap the entry's actual inet_address but
     // nonetheless the function can be used to promote the entry from
     // expiring to permanent or vice versa.
+    //
+    // Non-expiring entries can only be set on shard 0 and are replicated to other shards.
+    // Expiring entries are shard-local.
     void set(raft::server_id id, gms::inet_address addr, bool expiring) {
-        auto [it, emplaced] = _map.try_emplace(std::move(id), std::move(addr));
+        if (!expiring && this_shard_id() != 0) {
```

Comment:
I'd drop this check as well. Where we run group 0 is not the business of the map. invoke_on_others will work fine on non-0 shard.

Reply:
This is an issue of correctness.

As I explained in the PR cover letter, initially I was thinking of a solution where entries could be replicated from any shard. However, any solution satisfying this requirement that I could come up with turned out to be very complex.

Thus I simplified the requirement and said we will only replicate from shard 0 and only non-expiring entries. This PR does not correctly handle replication from other shards - for example, due to racing calls to `set` and `set_expiring_flag`, you could end up with entries that you intended to be expiring but they won't be, and you end up with "garbage" entries. Therefore I put these checks in place so we don't mistakenly do it.

Reply:
I don't see how correctness will be violated if you remove the check for shard id. I understand your reasons to drop expiring entries replication, but I don't see what could possibly go wrong if we simply drop the check - all the replication code will continue working fine.

Reply:
All code will continue working fine, until somebody starts calling `set` (with `expiring = false`) and `set_expiring_flag` concurrently on multiple shards. Then we'll end up with entries which should be expiring but are not, or vice-versa. The check is there to prevent incorrect use of the class. It's just a precondition check. It forces the user to serialize operations which set non-expiring entries or which change non-expiring entries to expiring ones through a single shard.

Reply:
Uhm, I see what you mean, basically we would have to keep an eye on other shards attempts to mark entries expiring. 
But shouldn't we then assert something like assert(expiring == false || this_shard_id() == 0)?

Reply:
The correct assert is `assert(expiring == true || this_shard_id() == 0)` - non-expiring entries can only be set on shard 0.

And this is exactly what I'm checking here: if the condition `expiring == true || this_shard_id() == 0` is false (i.e. `!expiring && this_shard_id() != 0`), then we call `on_internal_error` (which is like an assertion but won't crash the node in production, throw instead - AFAIR our policy is to use `on_internal_error` in places where one normally would use `assert`)

 Note that to turn a non-expiring entry to expiring one, you have to call `set_expiring_flag`, `set` won't change non-expiring entries to expiring ones. So it's fine to call `set` with `expiring == true` on all shards, even if the entry we're `set`ting is non-expiring.